ndk/src/midi.rs

```diff
@@ -0,0 +1,286 @@
+    ///
+    /// This should cause a receiver to discard any pending MIDI data it may have accumulated and
+    /// not processed.
+    pub fn send_flush(&self) -> Result<()> {
+        let result = unsafe { ffi::AMidiInputPort_sendFlush(self.as_ptr()) };
+        NdkMediaError::from_status(result)
+    }
+
+    pub fn send_with_timestamp(&self, buffer: &[u8], timestamp: i64) -> Result<usize> {
+        let num_bytes_sent = unsafe {
+            ffi::AMidiInputPort_sendWithTimestamp(
+                self.as_ptr(),
+                buffer.as_ptr(),
+                buffer.len() as ffi::size_t,
+                timestamp,
+            )
+        };
+        if num_bytes_sent >= 0 {
+            Ok(num_bytes_sent as usize)
+        } else {
+            Err(
+                NdkMediaError::from_status(ffi::media_status_t(num_bytes_sent as c_int))
+                    .unwrap_err(),
+            )
+        }
+    }
+}
+
+impl<'a> Drop for MidiInputPort<'a> {
+    fn drop(&mut self) {
+        unsafe { ffi::AMidiInputPort_close(self.as_ptr()) };
+    }
+}
+
+pub struct MidiOutputPort<'a> {
+    inner: NonNull<ffi::AMidiOutputPort>,
+    _marker: PhantomData<&'a MidiDevice>,
+}
+
+impl<'a> fmt::Debug for MidiOutputPort<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MidiOutputPort")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<'a> MidiOutputPort<'a> {
+    /// Creates an `MidiOutputPort` from a pointer.
+    ///
+    /// # Safety
+    /// By calling this function, you assert that the pointer is a valid pointer to a native
+    /// `AMidiOutputPort`.
+    pub unsafe fn from_ptr(inner: NonNull<ffi::AMidiOutputPort>) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+
+    fn as_ptr(&self) -> *mut ffi::AMidiOutputPort {
+        self.inner.as_ptr()
+    }
+
+    /// Receives the next pending MIDI message.
+    ///
+    /// To retrieve all pending messages, the client should repeatedly call this method until it
+    /// returns [`Ok(MidiOpcode::NoMessage)`].
+    ///
+    /// Note that this is a non-blocking call. If there are no Midi messages are available, the
+    /// function returns [`Ok(MidiOpcode::NoMessage)`] immediately (for 0 messages received).
+    pub fn receive(&self, buffer: &mut [u8]) -> Result<(MidiOpcode, i64)> {
+        let mut opcode = 0i32;
+        let mut timestamp = 0i64;
+        let mut num_bytes_received: ffi::size_t = 0;
+        let result = unsafe {
+            ffi::AMidiOutputPort_receive(
+                self.as_ptr(),
+                &mut opcode,
+                buffer.as_mut_ptr(),
+                buffer.len() as ffi::size_t,
+                &mut num_bytes_received,
+                &mut timestamp,
+            )
+        };
+
+        if result < 0 {
+            Err(NdkMediaError::from_status(ffi::media_status_t(result as c_int)).unwrap_err())
+        } else if result == 0 {
```

Comment:
The docs say that this function only ever returns 0 or 1 (and negative for error codes) - should we validate that it never returns anything `> 1` as we do in other functions with a `match` block?

Reply:
That's from the official sample:

https://github.com/android/ndk-samples/blob/7f6936ea044ee29c36b5c3ebd62bb3a64e1e6014/native-midi/app/src/main/cpp/AppMidiManager.cpp#L115-L129

AMidi always returns negative value for errors (`AMEDIA_*` enums are all negative except for `AMIDIA_OK`), I thought there's no possibility that `AMidiOutputPort_receive` is updated to return positive error values in future versions.

Reply:
Hmm.

We have a similar case for `fn pre_dispatch()` at #325, but there the docs don't specify whether non-boolean, non-error values are possible and it seems Android on certain devices has been tweaked to return these.

Reply:
If that's the case, should we return `Err(NdkMediaError::from_status(...))` as well as we did in `if result < 0 { ... }`?

Reply:
I was thinking something like:

```rust
match result {
    r if r < 0 => Err(NdkMediaError::from_status(ffi::media_status_t(r as c_int)).unwrap_err()),
    0 => Ok(MidiOpcode::NoMessage),
    1 => Ok(MidiOpcode::Data(num_bytes_received)),
    r => unreachable!("Result is positive integer {}", r),
}
```

Akin to https://github.com/rust-windowing/android-ndk-rs/blob/ddc2757f5abd002916c86576eaffd3f502e334b5/ndk/src/utils.rs#L8-L14

Reply:
Great! Since it seems that the original version `AMidiOutputPort_receive` definitely does not return any value but 0, 1, and `AMEDIA_ERROR_UNKNOWN`, I think this is better. I'll make a commit according to your suggestion.

Reply:
Fixed!